File: app-ios-nativa/Dominio/Pruebas/PruebasDominio/PruebasLogin.swift
import Foundation
import Testing

@testable import Dominio

@Suite("La vuelta del inicio de sesión")
struct PruebasLeerCallback {
    private func leer(_ texto: String) -> Login.Resultado {
        Login.leerCallback(URL(string: texto)!)
    }

    @Test("con código, es un inicio de sesión que fue bien")
    func conCodigo() {
        #expect(
            leer("guardalonativo://auth-callback?codigo=abc123")
                == .exito(codigoDeCanje: "abc123")
        )
    }

    @Test("sin correo, lo dice con sus palabras y no con el motivo del contrato")
    func sinEmail() {
        #expect(
            leer("guardalonativo://auth-callback?error=sin_email")
                == .error(
                    mensaje:
                        "Tu cuenta no ha dado ningún correo, y hace falta para crear la cuenta."
                )
        )
    }

    @Test("si el proveedor rechazó el intercambio, invita a repetir")
    func falloIntercambio() {
        #expect(
            leer("guardalonativo://auth-callback?error=fallo_intercambio")
                == .error(mensaje: "Google rechazó el inicio de sesión. Vuelve a intentarlo.")
        )
    }

    @Test("un motivo que no se conoce no deja al usuario sin explicación")
    func motivoDesconocido() {
        #expect(
            leer("guardalonativo://auth-callback?error=algo_nuevo")
                == .error(mensaje: "No se pudo iniciar sesión.")
        )
    }

    @Test("una vuelta sin nada tampoco se da por buena")
    func sinParametros() {
        #expect(
            leer("guardalonativo://auth-callback") == .error(mensaje: "No se pudo iniciar sesión."))
    }

    @Test("un código vacío no cuenta como haber entrado")
    func codigoVacio() {
        #expect(
            leer("guardalonativo://auth-callback?codigo=")
                == .error(mensaje: "No se pudo iniciar sesión."))
    }
}

@Suite("El estado del inicio de sesión")
struct PruebasEstadoLogin {
    @Test("son 32 caracteres hexadecimales")
    func formato() {
        let estado = Login.generarEstado()

        #expect(estado.count == 32)
        #expect(estado.allSatisfy { $0.isHexDigit })
    }

    @Test("dos seguidos no se repiten")
    func noSeRepite() {
        #expect(Login.generarEstado() != Login.generarEstado())
    }

    @Test("se puede fijar para probarlo")
    func inyectable() {
        #expect(
            Login.generarEstado(aleatorio: { cuantos in Array(repeating: 0xAB, count: cuantos) })
                == String(repeating: "ab", count: 16))
    }
}

@Suite("Textos de ajustes y de entrar")
struct PruebasTextosAjustes {
    @Test("el singular y el plural de los enlaces que ya había")
    func preguntaImportar() {
        #expect(
            Textos.preguntaImportar(cuantos: 1, deOtraCuenta: false)
                .hasPrefix("Hay 1 enlace guardado en este iPhone sin cuenta.")
        )
        #expect(
            Textos.preguntaImportar(cuantos: 3, deOtraCuenta: true)
                .hasPrefix("Hay 3 enlaces guardados en este iPhone con otra cuenta.")
        )
    }

    @Test("la pregunta dice que borrar no se puede deshacer")
    func avisaDeLoIrreversible() {
        #expect(
            Textos.preguntaImportar(cuantos: 2, deOtraCuenta: false).contains(
                "no se pueden recuperar"))
    }

    @Test("con sesión iniciada se dice con qué cuenta")
    func conCuenta() {
        #expect(
            Textos.sesionIniciadaComo(email: "persona@ejemplo.com")
                == "Sesión iniciada como persona@ejemplo.com. Tus enlaces se sincronizan con el PC."
        )
    }
}

File: app-android/dominio/src/test/kotlin/com/jmortizsilva/guardarenlaces/dominio/PruebasLogin.kt
package com.jmortizsilva.guardarenlaces.dominio

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

class PruebasLeerCallback {
    private fun leer(texto: String, proveedor: Proveedor = Proveedor.Google) =
        Login.leerCallback(texto, proveedor)

    @Test
    fun `con codigo, es un inicio de sesion que fue bien`() {
        assertEquals(
            Login.Resultado.Exito("abc123"),
            leer("guardarenlaces://auth-callback?codigo=abc123"),
        )
    }

    @Test
    fun `sin correo, lo dice con sus palabras y no con el motivo del contrato`() {
        assertEquals(
            Login.Resultado.Error(
                "Tu cuenta no ha dado ningún correo, y hace falta para crear la cuenta."
            ),
            leer("guardarenlaces://auth-callback?error=sin_email"),
        )
    }

    @Test
    fun `si el proveedor rechazo el intercambio, dice cual e invita a repetir`() {
        assertEquals(
            Login.Resultado.Error("Google rechazó el inicio de sesión. Vuelve a intentarlo."),
            leer("guardarenlaces://auth-callback?error=fallo_intercambio"),
        )
        // En iOS esto no puede pasar, porque Apple no va por web. Aquí sí, y decir «Google»
        // a quien estaba entrando con Apple es mandarle a mirar en el sitio equivocado.
        assertEquals(
            Login.Resultado.Error("Apple rechazó el inicio de sesión. Vuelve a intentarlo."),
            leer("guardarenlaces://auth-callback?error=fallo_intercambio", Proveedor.Apple),
        )
    }

    @Test
    fun `un motivo que no se conoce no deja al usuario sin explicacion`() {
        assertEquals(
            Login.Resultado.Error("No se pudo iniciar sesión."),
            leer("guardarenlaces://auth-callback?error=algo_nuevo"),
        )
    }

    @Test
    fun `una vuelta sin nada tampoco se da por buena`() {
        assertEquals(
            Login.Resultado.Error("No se pudo iniciar sesión."),
            leer("guardarenlaces://auth-callback"),
        )
    }

    @Test
    fun `un codigo vacio no cuenta como haber entrado`() {
        assertEquals(
            Login.Resultado.Error("No se pudo iniciar sesión."),
            leer("guardarenlaces://auth-callback?codigo="),
        )
    }

    @Test
    fun `si el codigo viene dos veces vale el primero`() {
        assertEquals(
            Login.Resultado.Exito("uno"),
            leer("guardarenlaces://auth-callback?codigo=uno&codigo=dos"),
        )
    }
}

class PruebasEstadoLogin {
    @Test
    fun `son 32 caracteres hexadecimales`() {
        val estado = Login.generarEstado()

        assertEquals(32, estado.length)
        assertTrue(estado.all { it in "0123456789abcdef" })
    }

    @Test
    fun `dos seguidos no se repiten`() {
        assertNotEquals(Login.generarEstado(), Login.generarEstado())
    }

    @Test
    fun `se puede fijar para probarlo`() {
        assertEquals(
            "ab".repeat(16),
            Login.generarEstado { cuantos -> ByteArray(cuantos) { 0xAB.toByte() } },
        )
    }
}
